In [1]:
import sqlite3
import numpy as np
import pandas as pd
%matplotlib notebook
import matplotlib.pyplot as plt
import xgboost as xgb
from xgboost.sklearn import XGBRegressor
from xgboost import plot_importance

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import train_test_split, GridSearchCV, ShuffleSplit, RandomizedSearchCV
from sklearn.pipeline import make_pipeline

import pickle

In [2]:
cnx = sqlite3.connect(r'../data/database.sqlite')

In [3]:
dd = pd.read_sql_query("SELECT name FROM sqlite_master WHERE type='table'", cnx)

In [4]:
print(dd)

                name
0    sqlite_sequence
1  Player_Attributes
2             Player
3              Match
4             League
5            Country
6               Team
7    Team_Attributes


In [5]:
df = pd.read_sql_query("SELECT * FROM Player_Attributes", cnx)

In [6]:
df.head()

id  player_fifa_api_id  player_api_id                 date  overall_rating  \
0   1              218353         505942  2016-02-18 00:00:00            67.0   
1   2              218353         505942  2015-11-19 00:00:00            67.0   
2   3              218353         505942  2015-09-21 00:00:00            62.0   
3   4              218353         505942  2015-03-20 00:00:00            61.0   
4   5              218353         505942  2007-02-22 00:00:00            61.0   

   potential preferred_foot attacking_work_rate defensive_work_rate  crossing  \
0       71.0          right              medium              medium      49.0   
1       71.0          right              medium              medium      49.0   
2       66.0          right              medium              medium      49.0   
3       65.0          right              medium              medium      48.0   
4       65.0          right              medium              medium      48.0   

   ...  vision  penalties  marking  standing_tackle  sliding_tackle  \
0  ...    54.0       48.0     65.0             69.0            69.0   
1  ...    54.0       48.0     65.0             69.0            69.0   
2  ...    54.0       48.0     65.0             66.0            69.0   
3  ...    53.0       47.0     62.0             63.0            66.0   
4  ...    53.0       47.0     62.0             63.0            66.0   

   gk_diving  gk_handling  gk_kicking  gk_positioning  gk_reflexes  
0        6.0         11.0        10.0             8.0          8.0  
1        6.0         11.0        10.0             8.0          8.0  
2        6.0         11.0        10.0             8.0          8.0  
3        5.0         10.0         9.0             7.0          7.0  
4        5.0         10.0         9.0             7.0          7.0  

[5 rows x 42 columns]

In [7]:
df.shape

(183978, 42)

In [8]:
target = df.pop('overall_rating')

In [9]:
type(target)

pandas.core.series.Series

In [10]:
df.shape

(183978, 41)

In [11]:
target.head()

0    67.0
1    67.0
2    62.0
3    61.0
4    61.0
Name: overall_rating, dtype: float64

## Imputing target funtion :

In [12]:
target.isnull().values.sum()

836

### there are 836 missing value present in target function.

In [13]:
target.describe()

count    183142.000000
mean         68.600015
std           7.041139
min          33.000000
25%          64.000000
50%          69.000000
75%          73.000000
max          94.000000
Name: overall_rating, dtype: float64

In [14]:
y = target.fillna(target.mean())

In [15]:
y.isnull().values.any()

False

## Data Exploration :

In [16]:
df.columns

Index(['id', 'player_fifa_api_id', 'player_api_id', 'date', 'potential',
       'preferred_foot', 'attacking_work_rate', 'defensive_work_rate',
       'crossing', 'finishing', 'heading_accuracy', 'short_passing', 'volleys',
       'dribbling', 'curve', 'free_kick_accuracy', 'long_passing',
       'ball_control', 'acceleration', 'sprint_speed', 'agility', 'reactions',
       'balance', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'aggression', 'interceptions', 'positioning', 'vision', 'penalties',
       'marking', 'standing_tackle', 'sliding_tackle', 'gk_diving',
       'gk_handling', 'gk_kicking', 'gk_positioning', 'gk_reflexes'],
      dtype='object')

In [17]:
for col in df.columns:
    unique_cat = len(df[col].unique())
    print("{col}--> {unique_cat}..{typ}".format(col=col, unique_cat=unique_cat, typ=df[col].dtype))

id--> 183978..int64
player_fifa_api_id--> 11062..int64
player_api_id--> 11060..int64
date--> 197..object
potential--> 57..float64
preferred_foot--> 3..object
attacking_work_rate--> 9..object
defensive_work_rate--> 20..object
crossing--> 96..float64
finishing--> 98..float64
heading_accuracy--> 97..float64
short_passing--> 96..float64
volleys--> 94..float64
dribbling--> 98..float64
curve--> 93..float64
free_kick_accuracy--> 98..float64
long_passing--> 96..float64
ball_control--> 94..float64
acceleration--> 87..float64
sprint_speed--> 86..float64
agility--> 82..float64
reactions--> 79..float64
balance--> 82..float64
shot_power--> 97..float64
jumping--> 80..float64
stamina--> 85..float64
strength--> 83..float64
long_shots--> 97..float64
aggression--> 92..float64
interceptions--> 97..float64
positioning--> 96..float64
vision--> 98..float64
penalties--> 95..float64
marking--> 96..float64
standing_tackle--> 96..float64
sliding_tackle--> 95..float64
gk_diving--> 94..float64
gk_handling--> 91..

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183978 entries, 0 to 183977
Data columns (total 41 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   id                   183978 non-null  int64  
 1   player_fifa_api_id   183978 non-null  int64  
 2   player_api_id        183978 non-null  int64  
 3   date                 183978 non-null  object 
 4   potential            183142 non-null  float64
 5   preferred_foot       183142 non-null  object 
 6   attacking_work_rate  180748 non-null  object 
 7   defensive_work_rate  183142 non-null  object 
 8   crossing             183142 non-null  float64
 9   finishing            183142 non-null  float64
 10  heading_accuracy     183142 non-null  float64
 11  short_passing        183142 non-null  float64
 12  volleys              181265 non-null  float64
 13  dribbling            183142 non-null  float64
 14  curve                181265 non-null  float64
 15  free_kick_accurac

In [19]:
df.isnull().sum()

id                        0
player_fifa_api_id        0
player_api_id             0
date                      0
potential               836
preferred_foot          836
attacking_work_rate    3230
defensive_work_rate     836
crossing                836
finishing               836
heading_accuracy        836
short_passing           836
volleys                2713
dribbling               836
curve                  2713
free_kick_accuracy      836
long_passing            836
ball_control            836
acceleration            836
sprint_speed            836
agility                2713
reactions               836
balance                2713
shot_power              836
jumping                2713
stamina                 836
strength                836
long_shots              836
aggression              836
interceptions           836
positioning             836
vision                 2713
penalties               836
marking                 836
standing_tackle         836
sliding_tackle      

In [20]:
df.describe().T

count           mean            std     min        25%  \
id                  183978.0   91989.500000   53110.018250     1.0   45995.25   
player_fifa_api_id  183978.0  165671.524291   53851.094769     2.0  155798.00   
player_api_id       183978.0  135900.617324  136927.840510  2625.0   34763.00   
potential           183142.0      73.460353       6.592271    39.0      69.00   
crossing            183142.0      55.086883      17.242135     1.0      45.00   
finishing           183142.0      49.921078      19.038705     1.0      34.00   
heading_accuracy    183142.0      57.266023      16.488905     1.0      49.00   
short_passing       183142.0      62.429672      14.194068     3.0      57.00   
volleys             181265.0      49.468436      18.256618     1.0      35.00   
dribbling           183142.0      59.175154      17.744688     1.0      52.00   
curve               181265.0      52.965675      18.255788     2.0      41.00   
free_kick_accuracy  183142.0      49.380950      17.831746     1.0      36.00   
long_passing        183142.0      57.069880      14.394464     3.0      49.00   
ball_control        183142.0      63.388879      15.196671     5.0      58.00   
acceleration        183142.0      67.659357      12.983326    10.0      61.00   
sprint_speed        183142.0      68.051244      12.569721    12.0      62.00   
agility             181265.0      65.970910      12.954585    11.0      58.00   
reactions           183142.0      66.103706       9.155408    17.0      61.00   
balance             181265.0      65.189496      13.063188    12.0      58.00   
shot_power          183142.0      61.808427      16.135143     2.0      54.00   
jumping             181265.0      66.969045      11.006734    14.0      60.00   
stamina             183142.0      67.038544      13.165262    10.0      61.00   
strength            183142.0      67.424529      12.072280    10.0      60.00   
long_shots          183142.0      53.339431      18.367025     1.0      41.00   
aggression          183142.0      60.948046      16.089521     6.0      51.00   
interceptions       183142.0      52.009271      19.450133     1.0      34.00   
positioning         183142.0      55.786504      18.448292     2.0      45.00   
vision              181265.0      57.873550      15.144086     1.0      49.00   
penalties           183142.0      55.003986      15.546519     2.0      45.00   
marking             183142.0      46.772242      21.227667     1.0      25.00   
standing_tackle     183142.0      50.351257      21.483706     1.0      29.00   
sliding_tackle      181265.0      48.001462      21.598778     2.0      25.00   
gk_diving           183142.0      14.704393      16.865467     1.0       7.00   
gk_handling         183142.0      16.063612      15.867382     1.0       8.00   
gk_kicking          183142.0      20.998362      21.452980     1.0       8.00   
gk_positioning      183142.0      16.132154      16.099175     1.0       8.00   
gk_reflexes         183142.0      16.441439      17.198155     1.0       8.00   

                         50%        75%       max  
id                   91989.5  137983.75  183978.0  
player_fifa_api_id  183488.0  199848.00  234141.0  
player_api_id        77741.0  191080.00  750584.0  
potential               74.0      78.00      97.0  
crossing                59.0      68.00      95.0  
finishing               53.0      65.00      97.0  
heading_accuracy        60.0      68.00      98.0  
short_passing           65.0      72.00      97.0  
volleys                 52.0      64.00      93.0  
dribbling               64.0      72.00      97.0  
curve                   56.0      67.00      94.0  
free_kick_accuracy      50.0      63.00      97.0  
long_passing            59.0      67.00      97.0  
ball_control            67.0      73.00      97.0  
acceleration            69.0      77.00      97.0  
sprint_speed            69.0      77.00      97.0  
agility                 68.0      75.00      96.0  
reaction

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183978 entries, 0 to 183977
Data columns (total 41 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   id                   183978 non-null  int64  
 1   player_fifa_api_id   183978 non-null  int64  
 2   player_api_id        183978 non-null  int64  
 3   date                 183978 non-null  object 
 4   potential            183142 non-null  float64
 5   preferred_foot       183142 non-null  object 
 6   attacking_work_rate  180748 non-null  object 
 7   defensive_work_rate  183142 non-null  object 
 8   crossing             183142 non-null  float64
 9   finishing            183142 non-null  float64
 10  heading_accuracy     183142 non-null  float64
 11  short_passing        183142 non-null  float64
 12  volleys              181265 non-null  float64
 13  dribbling            183142 non-null  float64
 14  curve                181265 non-null  float64
 15  free_kick_accurac

In [22]:
df.isnull().sum()

id                        0
player_fifa_api_id        0
player_api_id             0
date                      0
potential               836
preferred_foot          836
attacking_work_rate    3230
defensive_work_rate     836
crossing                836
finishing               836
heading_accuracy        836
short_passing           836
volleys                2713
dribbling               836
curve                  2713
free_kick_accuracy      836
long_passing            836
ball_control            836
acceleration            836
sprint_speed            836
agility                2713
reactions               836
balance                2713
shot_power              836
jumping                2713
stamina                 836
strength                836
long_shots              836
aggression              836
interceptions           836
positioning             836
vision                 2713
penalties               836
marking                 836
standing_tackle         836
sliding_tackle      

In [23]:
## Seprate numerical and categorical features
numerical_columns=df.columns[df.dtypes!='object']
categorical_columns=df.columns[df.dtypes=='object']
print("Numerical columns:",numerical_columns)
print('Categorical Columns:',categorical_columns)

Numerical columns: Index(['id', 'player_fifa_api_id', 'player_api_id', 'potential', 'crossing',
       'finishing', 'heading_accuracy', 'short_passing', 'volleys',
       'dribbling', 'curve', 'free_kick_accuracy', 'long_passing',
       'ball_control', 'acceleration', 'sprint_speed', 'agility', 'reactions',
       'balance', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'aggression', 'interceptions', 'positioning', 'vision', 'penalties',
       'marking', 'standing_tackle', 'sliding_tackle', 'gk_diving',
       'gk_handling', 'gk_kicking', 'gk_positioning', 'gk_reflexes'],
      dtype='object')
Categorical Columns: Index(['date', 'preferred_foot', 'attacking_work_rate', 'defensive_work_rate'], dtype='object')


### we can see only four features have the type 'object'. here the feature named 'date' has no significance in this problem so can ignore it and perform one hot encoding on the rest of 3 features.

In [24]:
dummy_df = pd.get_dummies(df, columns=['preferred_foot', 'attacking_work_rate', 'defensive_work_rate'])
dummy_df.head()

id  player_fifa_api_id  player_api_id                 date  potential  \
0   1              218353         505942  2016-02-18 00:00:00       71.0   
1   2              218353         505942  2015-11-19 00:00:00       71.0   
2   3              218353         505942  2015-09-21 00:00:00       66.0   
3   4              218353         505942  2015-03-20 00:00:00       65.0   
4   5              218353         505942  2007-02-22 00:00:00       65.0   

   crossing  finishing  heading_accuracy  short_passing  volleys  ...  \
0      49.0       44.0              71.0           61.0     44.0  ...   
1      49.0       44.0              71.0           61.0     44.0  ...   
2      49.0       44.0              71.0           61.0     44.0  ...   
3      48.0       43.0              70.0           60.0     43.0  ...   
4      48.0       43.0              70.0           60.0     43.0  ...   

   defensive_work_rate_9  defensive_work_rate__0  defensive_work_rate_ean  \
0                  False                   False                    False   
1                  False                   False                    False   
2                  False                   False                    False   
3                  False                   False                    False   
4                  False                   False                    False   

   defensive_work_rate_es  defensive_work_rate_high  defensive_work_rate_low  \
0                   False                     False                    False   
1                   False                     False                    False   
2                   False                     False                    False   
3                   False                     False                    False   
4                   False                     False                    False   

   defensive_work_rate_medium  defensive_work_rate_o  \
0                        True                  False   
1                        True                  False   
2                        True                  False   
3                        True                  False   
4                        True                  False   

   defensive_work_rate_ormal  defensive_work_rate_tocky  
0                      False                      False  
1                      False                      False  
2                      False                      False  
3                      False                      False  
4                      False                      False  

[5 rows x 67 columns]

In [25]:
X = dummy_df.drop(['id', 'date'], axis=1)

## Feature selection :

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [28]:
from sklearn.impute import SimpleImputer as Imputer
#imputing null value of each column with the mean of that column
imput = Imputer()
X_train = imput.fit_transform(X_train)
X_test = imput.fit_transform(X_test)

In [29]:
#finding feature_importance for feature selection. from it we'll be able to decide threshold value
model = XGBRegressor()
model.fit(X_train, y_train)
print(model.feature_importances_)

[1.49139501e-02 1.42232142e-02 1.55819103e-01 1.29311280e-02
 8.96300375e-03 2.67720707e-02 2.99981348e-02 2.18823529e-03
 1.26897143e-02 1.02950400e-03 8.46192241e-04 3.82854184e-03
 6.07885309e-02 3.75495385e-03 8.57065897e-03 1.38577376e-03
 3.16335618e-01 9.92643530e-04 1.71729811e-02 2.13329750e-03
 5.42683294e-03 1.41720586e-02 4.75015119e-03 8.62127636e-03
 1.06486725e-02 1.74707025e-02 2.52074655e-03 1.27184228e-03
 3.17236409e-02 5.59478551e-02 5.20201353e-03 3.64738405e-02
 2.27714442e-02 1.00484760e-02 3.70792300e-02 2.17788834e-02
 1.01251272e-03 0.00000000e+00 8.63931142e-04 8.47549760e-04
 8.24667746e-04 5.72213496e-04 9.41885402e-04 1.46208378e-03
 0.00000000e+00 1.27837135e-04 0.00000000e+00 0.00000000e+00
 1.35666924e-03 0.00000000e+00 8.71833239e-04 0.00000000e+00
 2.60954746e-03 0.00000000e+00 0.00000000e+00 0.00000000e+00
 0.00000000e+00 0.00000000e+00 0.00000000e+00 9.94736329e-04
 4.14713693e-04 5.08083496e-04 5.34691522e-03 0.00000000e+00
 0.00000000e+00]


In [30]:
selection = SelectFromModel(model, threshold=0.01, prefit=True)

select_X_train = selection.transform(X_train)
select_X_test = selection.transform(X_test)

In [31]:
select_X_train.shape

(137983, 20)

## Training different models :

1. Linear Regression :

In [32]:
pipe = make_pipeline(StandardScaler(),             #preprocessing(standard scalling)
                     LinearRegression())           #estimator(linear regression)

param_grid = {'linearregression__n_jobs': [-1]}     #parameters for model tunning

grid = GridSearchCV(pipe, param_grid=param_grid, cv=10)

In [33]:
grid.fit(select_X_train, y_train)          #training 

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('standardscaler', StandardScaler()),
                                       ('linearregression',
                                        LinearRegression())]),
             param_grid={'linearregression__n_jobs': [-1]})

In [34]:
grid.best_params_

{'linearregression__n_jobs': -1}

In [35]:
lin_reg = pickle.dumps(grid)

2. Decision Tree :

In [41]:
pipe = make_pipeline(StandardScaler(),                  #preprocessing
                     DecisionTreeRegressor(criterion='squared_error', random_state=0))          #estimator

param_grid = {'decisiontreeregressor__max_depth': [3, 5, 9, 13]}

grid = GridSearchCV(pipe, param_grid=param_grid, cv=10)

In [42]:
grid.fit(select_X_train, y_train)          #training 

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('standardscaler', StandardScaler()),
                                       ('decisiontreeregressor',
                                        DecisionTreeRegressor(random_state=0))]),
             param_grid={'decisiontreeregressor__max_depth': [3, 5, 9, 13]})

In [43]:
grid.best_params_

{'decisiontreeregressor__max_depth': 13}

In [44]:
Dectree_reg = pickle.dumps(grid)

3. Ranom Forest :

In [45]:
pipe = make_pipeline(StandardScaler(),
                     RandomForestRegressor(n_estimators=200, random_state=0))

#cv = ShuffleSplit(test_size=0.2, random_state=0)

param_grid = {'randomforestregressor__max_features':['sqrt', 'log2'],
              'randomforestregressor__max_depth':[9, 10]}                 

grid = GridSearchCV(pipe, param_grid=param_grid, cv=5)

In [46]:
grid.fit(select_X_train, y_train)          #training 

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('standardscaler', StandardScaler()),
                                       ('randomforestregressor',
                                        RandomForestRegressor(n_estimators=200,
                                                              random_state=0))]),
             param_grid={'randomforestregressor__max_depth': [9, 10],
                         'randomforestregressor__max_features': ['sqrt',
                                                                 'log2']})

In [47]:
grid.best_params_

{'randomforestregressor__max_depth': 10,
 'randomforestregressor__max_features': 'sqrt'}

In [48]:
Randfor_reg = pickle.dumps(grid)

4. Xgboost regressor :

In [49]:
pipe = make_pipeline(StandardScaler(),
                     XGBRegressor(n_estimators=200, random_state=42))

#cv = ShuffleSplit(n_splits=10, random_state=0)

param_grid = {'xgbregressor__max_depth': [5, 7],
              'xgbregressor__learning_rate': [0.1, 0.3]}

grid = GridSearchCV(pipe, param_grid=param_grid, cv=5, n_jobs= -1)

In [50]:
grid.fit(select_X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('standardscaler', StandardScaler()),
                                       ('xgbregressor',
                                        XGBRegressor(base_score=None,
                                                     booster=None,
                                                     callbacks=None,
                                                     colsample_bylevel=None,
                                                     colsample_bynode=None,
                                                     colsample_bytree=None,
                                                     device=None,
                                                     early_stopping_rounds=None,
                                                     enable_categorical=False,
                                                     eval_metric=None,
                                                     feature_types=None,
                                                     gamma=None,
                                                     grow_policy=None,
                                                     im...
                                                     max_bin=None,
                                                     max_cat_threshold=None,
                                                     max_cat_to_onehot=None,
                                                     max_delta_step=None,
                                                     max_depth=None,
                                                     max_leaves=None,
                                                     min_child_weight=None,
                                                     missing=nan,
                                                     monotone_constraints=None,
                                                     multi_strategy=None,
                                                     n_estimators=200,
                                                     n_jobs=None,
                                                     num_parallel_tree=None,
                                                     random_state=42, ...))]),
             n_jobs=-1,
             param_grid={'xgbregressor__learning_rate': [0.1, 0.3],
                         'xgbregressor__max_depth': [5, 7]})

In [51]:
grid.best_params_

{'xgbregressor__learning_rate': 0.3, 'xgbregressor__max_depth': 7}

In [52]:
xgbreg = pickle.dumps(grid)

### Comparision between different models :

In [53]:
lin_reg = pickle.loads(lin_reg)
Dectree_reg = pickle.loads(Dectree_reg)
Randfor_reg = pickle.loads(Randfor_reg)
xgbreg = pickle.loads(xgbreg)

In [54]:
print("""Linear Regressor accuracy is {lin}
DecisionTree Regressor accuracy is {Dec}
RandomForest regressor accuracy is {ran}
XGBoost regressor accuracy is {xgb}""".format(lin=lin_reg.score(select_X_test, y_test),
                                                       Dec=Dectree_reg.score(select_X_test, y_test),
                                                       ran=Randfor_reg.score(select_X_test, y_test),
                                                       xgb=xgbreg.score(select_X_test, y_test)))

Linear Regressor accuracy is 0.8534284367544658
DecisionTree Regressor accuracy is 0.9279688664805696
RandomForest regressor accuracy is 0.9386216691432722
XGBoost regressor accuracy is 0.9692021703247189
